firstvoices/backend/views/custom_order_recalculate_views.py

```diff
@@ -54,3 +63,55 @@ def create(self, request, *args, **kwargs):
 
         except recalculate_custom_order_preview.OperationalError:
             raise CeleryError()
+
+
+class CustomOrderRecalculateView(APIView):
+    tasks = []
+
+    @staticmethod
+    def check_permission(request):
+        if not rules.has_perm("views.has_custom_order_access", request.user):
+            raise PermissionDenied()
+
+    def get(self, request, site_slug: str):
+        # Check Superadmin status
+        self.check_permission(request)
+
+        # Check that the site exists
+        try:
+            Site.objects.get(slug=site_slug)
+        except ObjectDoesNotExist:
+            raise Http404
+
+        # Return the status of any ongoing recalculation task
+        for task in self.tasks:
+            if task["site_slug"] == site_slug:
+                async_result = AsyncResult(task["task_id"])
+                return Response(
+                    {"current_recalculation_task_status": async_result.status}
+                )
+        return Response({"current_recalculation_task_status": "Not started."})
+
+    def post(self, request, site_slug: str):
+        # Check Superadmin status
+        self.check_permission(request)
+
+        # Check that the site exists
+        try:
+            Site.objects.get(slug=site_slug)
+        except ObjectDoesNotExist:
+            raise Http404
+
+        # Call the recalculation task
+        try:
+            recalculation_results = recalculate_custom_order.apply_async((site_slug,))
+            task = {"task_id": recalculation_results.task_id, "site_slug": site_slug}
+            self.tasks.append(task)
+
+            return Response(
+                {"recalculation_results": recalculation_results.get(timeout=360)},
+                status=200,
```

Comment:
Wait, how does this return the final result if the task is running asynchronously?

Reply:
I have the same question. Unless I am missing something it looks like if the recalculation takes longer than the timeout it will raise an error?